src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestCallback.cs

```diff
@@ -305,6 +305,16 @@ private static void OnRequestError(WinHttpRequestState state, Interop.WinHttp.WI
             
             Debug.Assert(state != null, "OnRequestError: state is null");
 
+            // Add instrumentation for now to catch unexpected errors and fail fast and produce a crash dump.
+            // Issue #7812
+#if DEBUG
+            if (unchecked((int)asyncResult.dwError) == Interop.WinHttp.ERROR_INSUFFICIENT_BUFFER ||
+                unchecked((int)asyncResult.dwError) == unchecked((int)0x80090321)) // SEC_E_BUFFER_TOO_SMALL
+            {
+                Debug.Assert(false, $"Unexpected error: {unchecked((int)asyncResult.dwError)}, WinHttp API: {unchecked((uint)asyncResult.dwResult.ToInt32())}");
```

Comment:
Debug.Assert has the documentation that "outputs a specified message and displays a message box that shows the call stack", while Debug.Fail has "Emits the specified error message". Based on documentation, I prefer Debug.Assert.